### PR TITLE
feat: add `api_method` parameter to `Client.query` to select `INSERT`  or `QUERY` API

### LINKED ISSUE #589: Provide support for synchronous queries through the v2/projects/{projectId}/queries endpoint
**Is your feature request related to a problem? Please describe.**
We are part of the BI Engine preview so we're trying to minimize the latency between our python application and BigQuery. The current approach of creating an asynchronous job and then waiting for it to complete adds significant latency and hurts the value proposition of BI Engine.
 **Describe the solution you'd like**
We'd like a way to utilize the `v2/projects/{projectId}/queries` endpoint to execute a query and wait for the response.
 **Describe alternatives you've considered**
We could call the REST API directly. The main drawbacks are 1) authentication with a service account 2) the existing python client does a nice job of creating `Row` objects with the results.
I've also used the following snippet:
```python
bq_client = bigquery.Client(project=project_id)
synchronous_api_response = bq_client._connection.api_request(
    "POST", 
    f"/projects/{project_id}/queries",
    data={"query": query, "useLegacySql": False},
)
```

### PR DESCRIPTION
**Pull request by tswast**
This PR is the first step in enabling the "fast query path". At the moment, selecting `api_method="QUERY"` is actually likely to be slower because it fetches the first page of results and discards them. A PR will follow-up to cache this first page for use by the row iterator.

~This commit only refactors to allow jobs.insert to be selected.
Supporting jobs.query will require more transformations to QueryJobConfig,
QueryJob, and RowIterator.~

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Towards #589 🦕

### CODE REVIEW

File: google/cloud/bigquery/_job_helpers.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    request_body.update(job_config_resource["configuration"]["query"])
+    request_body["location"] = location
+    request_body["labels"] = job_config.labels
+    request_body["dryRun"] = job_config.dry_run
+
+    # Subtract a buffer for context switching, network latency, etc.
+    request_body["timeoutMs"] = max(0, int(1000 * (timeout - _TIMEOUT_BUFFER_SECS)))
+
+    def do_query():
+        request_body["requestId"] = make_job_id(None)
```
Comment (tswast) [inline google/cloud/bigquery/_job_helpers.py:141]: I think this should actually be outside the `do_query()` method. The behavior of `requestId` is a bit different from `jobId`.

**Comment by tswast** [comment]
Even though this isn't a breaking change, it refactors enough that I think this should target `v3` so as not to delay `v3` any longer.

**Comment by tswast** [comment]
I've just synced to the latest v3 branch. Unit tests and query system tests are passing. I think this is ready for review.

Note: This won't actually result in any performance improvements until the `# TODO: https://github.com/googleapis/python-bigquery/issues/589` in `_job_helpers.py` is resolved. Since that requires more hacks to the `QueryJob` class, I figure that's worth postponing to a separate PR. As-is, this PR should be showing the same behavior for `api_method="INSERT"` and `api_method="QUERY"` options.

**Comment by tswast** [comment]
Unit test failures on Python 3.6. Not sure if it's real or has to do with how we're mocking some objects.

```
                else:
                    reductor = dispatch_table.get(cls)
                    if reductor:
                        rv = reductor(x)
                    else:
                        reductor = getattr(x, "__reduce_ex__", None)
                        if reductor:
>                           rv = reductor(4)
E                           TypeError: can't pickle property objects

/usr/local/lib/python3.6/copy.py:169: TypeError
```

**Comment by tswast** [comment]
System test failures:

```
_ ERROR at teardown of test_structs[struct<name string(22), children array<struct<name string(22), bdate date>>>] _

bigquery_client = <google.cloud.bigquery.client.Client object at 0x7ffb8c7b4640>
project_id = 'precise-truck-742'
dataset_id = 'python_bigquery_tests_system_20211118230426_5b4dcd'

    @pytest.fixture(scope="session")
    def scalars_extreme_table(
        bigquery_client: bigquery.Client, project_id: str, dataset_id: str
    ):
        full_table_id = load_scalars_table(
            bigquery_client, project_id, dataset_id, data_path="scalars_extreme.jsonl"
        )
        yield full_table_id
>       bigquery_client.delete_table(full_table_id)

tests/system/conftest.py:136:
```

I think I need to add the "not found OK" parameter to `delete_table` in the fixture.

File: tests/system/conftest.py [hunk 2]
```diff
@@ -78,18 +90,37 @@ def table_id(dataset_id):
...
+    bigquery_client: bigquery.Client,
+    project_id: str,
+    dataset_id: str,
+    data_path: str = "scalars.jsonl",
+) -> str:
     schema = bigquery_client.schema_from_json(DATA_DIR / "scalars_schema.json")
     job_config = bigquery.LoadJobConfig()
     job_config.schema = schema
     job_config.source_format = enums.SourceFormat.NEWLINE_DELIMITED_JSON
     full_table_id = f"{project_id}.{dataset_id}.scalars"
```
Comment (tswast) [inline tests/system/conftest.py:103]: Need random/custom table ID to prevent conflicts with "extremes" and regular table.

**Comment by tswast** [comment]
Test coverage failures:

```
google/cloud/bigquery/client.py                                823      3    312      3    99%   3251, 3285, 3302
```

Also typing issues:

```
nox > mypy google/cloud
google/cloud/bigquery/client.py:3292: error: Argument 3 to "query_jobs_insert" has incompatible type "Optional[QueryJobConfig]"; expected "QueryJobConfig"
google/cloud/bigquery/client.py:3293: error: Argument 4 to "query_jobs_insert" has incompatible type "Optional[str]"; expected "str"
google/cloud/bigquery/client.py:3294: error: Argument 5 to "query_jobs_insert" has incompatible type "Optional[str]"; expected "str"
Found 3 errors in 1 file (checked 50 source files)
```

```
FAILED: /tmpfs/src/github/python-bigquery/.pytype/pyi/google/cloud/bigquery/_job_helpers.pyi
/tmpfs/src/github/python-bigquery/.nox/pytype/bin/python -m pytype.single --disable pyi-error --imports_info /tmpfs/src/github/python-bigquery/.pytype/imports/google.cloud.bigquery._job_helpers.imports --module-name google.cloud.bigquery._job_helpers -V 3.8 -o /tmpfs/src/github/python-bigquery/.pytype/pyi/google/cloud/bigquery/_job_helpers.pyi --analyze-annotated --nofail --quick /tmpfs/src/github/python-bigquery/google/cloud/bigquery/_job_helpers.py
File "/tmpfs/src/github/python-bigquery/google/cloud/bigquery/_job_helpers.py", line 141, in _to_query_request: unsupported operand type(s) for item assignment: 'bool' [unsupported-operands]
  No attribute '__setitem__' on bool

For more details, see https://google.github.io/pytype/errors.html#unsupported-operands
```

File: tests/system/test_query.py [hunk 3]
```diff
@@ -12,15 +12,430 @@
...
+def table_with_9999_columns_10_rows(bigquery_client, project_id, dataset_id):
+    """Generate a table of maximum width via CREATE TABLE AS SELECT.
+
+    The first column is named 'rowval', and has a value from 1..rowcount
+    Subsequent columns are named col_<N> and contain the value N*rowval, where
+    N is between 1 and 9999 inclusive.
+    """
+    table_id = "many_columns"
+    row_count = 10
+    col_projections = ",".join([f"r * {n} as col_{n}" for n in range(1, 10000)])
```
Comment (plamut) [inline tests/system/test_query.py:46]: (nit) Intermittent list is unnecessary.

File: tests/system/test_query.py [hunk 4]
```diff
@@ -29,15 +444,35 @@ def test_dry_run(bigquery_client: bigquery.Client, scalars_table: str):
...
+    with pytest.raises(exceptions.NotFound, match="not_a_real_dataset"):
+        bigquery_client.query(
+            "SELECT * FROM not_a_real_dataset.doesnt_exist", api_method="QUERY"
+        )
+
+
+def test_query_error_w_api_method_insert(bigquery_client: bigquery.Client):
+    """With jobs.insert, an exception is thrown when fetching the results.."""
+
+    query_job = bigquery_client.query("SELECT * FROM not_a_real_dataset.doesnt_exist")
```
Comment (plamut) [inline tests/system/test_query.py:459]: Maybe rename the test case to explain that we also test that "INSERT" is indeed the default?

File: tests/unit/test__job_helpers.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+                        },
+                        {
+                            "parameterType": {"type": "INT64"},
+                            "parameterValue": {"value": "123"},
+                        },
+                    ],
+                }
+            ),
+        ),
+        # TODO: connection properties
```
Comment (plamut) [inline tests/unit/test__job_helpers.py:127]: In this or in a separate PR?

**Review by plamut — COMMENTED** [review]
Looks really promising and structured, mostly just minor remarks and questions.

File: tests/unit/test__job_helpers.py (hunk 5 above)
Comment (tswast) [reply tests/unit/test__job_helpers.py:127]: Oops this PR. I created https://github.com/googleapis/python-bigquery/pull/1024 shortly after writing this TODO.

File: tests/system/test_query.py (hunk 4 above)
Comment (tswast) [reply tests/system/test_query.py:464]: Renamed and added more explanation to the comment that this (error only on `result()`) is the "old" behavior and is maintained for backwards compatibility.

File: google/cloud/bigquery/_job_helpers.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+from google.api_core import retry as retries
+
+from google.cloud.bigquery import job
+
+# Avoid circular imports
+if TYPE_CHECKING:  # pragma: NO COVER
+    from google.cloud.bigquery.client import Client
+
+
+_TIMEOUT_BUFFER_MILLIS = 100
```
[the remark this answers is not in the document]
Comment (tswast) [reply google/cloud/bigquery/_job_helpers.py]: Probably not, but also doesn't matter too much. I added some comments to the file with my thoughts.

File: google/cloud/bigquery/client.py [hunk 7]
```diff
@@ -3215,6 +3219,11 @@ def query(
                 called on the job returned. The ``job_retry``
                 specified here becomes the default ``job_retry`` for
                 ``result()``, where it can also be specified.
+            api_method (Union[str, enums.QueryApiMethod]):
```
Comment (plamut) [inline google/cloud/bigquery/client.py:3222]: (nit) Duplicate annotation info

**Review by plamut — APPROVED** [review]
LGTM.

There's one nit, but decide for yourself if it's worth updating the PR.

**Review by loferris — APPROVED** [review]
LGTM

**Pull request merged**